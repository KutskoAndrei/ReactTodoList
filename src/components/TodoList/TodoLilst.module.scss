.todoList {
  margin: 12px;
  padding: 8px;
  list-style-type: none;
}