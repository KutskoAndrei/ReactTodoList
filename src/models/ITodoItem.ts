interface ITodoItem {
    id: string;
    description: string;
    checked: boolean;
};

export default ITodoItem;