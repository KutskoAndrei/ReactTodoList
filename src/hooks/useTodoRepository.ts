import { ITodoItem } from "../models/ITodoItem";

export function useTodoRepository() {
    const key = 'todoList';

    function saveTodoList(todos: ITodoItem[]) {
        localStorage.setItem(key, JSON.stringify(todos));
    }

    function getTodoList(): ITodoItem[] {
        const data = localStorage.getItem(key);
        if(data) {
            const todos = JSON.parse(data) as ITodoItem[];
            return todos;
        }
        return [];
    }

    function getFinishedTodos(): ITodoItem[] {
        const data = localStorage.getItem(key);
        if(data) {
            const todos = JSON.parse(data) as ITodoItem[];
            return todos.filter((todo) => todo.checked);
        }
        return [];
    }

    function getUnfinishedTodos(): ITodoItem[] {
        const data = localStorage.getItem(key);
        if(data) {
            const todos = JSON.parse(data) as ITodoItem[];
            return todos.filter((todo) => !todo.checked);
        }
        return [];
    }

    function removeFinishedTodos(): ITodoItem[] {
        const todos = getUnfinishedTodos();
        saveTodoList(todos);
        return todos;
    }

    function switchTodo(todo: ITodoItem): ITodoItem[] {
        const data = localStorage.getItem(key);
        if(data) {
            const todos = JSON.parse(data) as ITodoItem[];
            const todoToSwitch = todos.find((item) => item.id === todo.id);
            if(todoToSwitch) {
                todoToSwitch.checked = !todoToSwitch.checked;
                saveTodoList(todos);
                return todos;
            }
        }
        return [];
    }

    return {
        saveTodoList,
        getTodoList,
        getFinishedTodos,
        getUnfinishedTodos,
        removeFinishedTodos,
        switchTodo
    }
}