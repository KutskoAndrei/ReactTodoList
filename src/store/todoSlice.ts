import { FilterType } from "../models/FilterType";
import { ITodoItem } from "../models/ITodoItem";
import { createSlice, PayloadAction } from '@reduxjs/toolkit'

const todos: ITodoItem[] = [];
const filter: number = FilterType.ALL;

const todoSlice = createSlice({
    name: 'todos',
    initialState: {
        todos,
        filter
    },
    reducers: {
        addTodo(state, action: PayloadAction<ITodoItem>) {
            state.todos.push(action.payload);
        },
        removeTodo(state, action: PayloadAction<ITodoItem>) {
            state.todos = state.todos.filter((todo) => todo.id !== action.payload.id);
        },
        switchTodoStatus(state, action: PayloadAction<ITodoItem>) {
            const switchTodo = state.todos.find((todo) => todo.id = action.payload.id);
            if (switchTodo) switchTodo.checked = !switchTodo.checked;
        },
        updateTodos(state, action: PayloadAction<ITodoItem[]>) {
            state.todos = action.payload;
        },
        updateFilter(state, action: PayloadAction<FilterType>) {
            state.filter = action.payload;
        }
    }
});

export const {
    addTodo,
    removeTodo,
    switchTodoStatus,
    updateTodos,
    updateFilter
} = todoSlice.actions;

export default todoSlice.reducer;