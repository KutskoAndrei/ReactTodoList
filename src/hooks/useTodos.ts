import FilterType from "../models/FilterType";
import ITodoItem from "../models/ITodoItem";
import { removeTodo, updateTodos } from "../store/todoSlice";
import { useAppDispatch } from "./useAppDispatch";
import { useAppSelector } from "./useAppSelector";
import { useTodoRepository } from "./useTodoRepository";

export function useTodos() {
    const {
        saveTodoList,
        getTodoList,
        getFinishedTodos,
        getUnfinishedTodos,
        removeFinishedTodos,
        switchTodo
    } = useTodoRepository();
    const dispatch = useAppDispatch();
    const filter = useAppSelector((state) => state.todos.filter);

    function set(todos: ITodoItem[]) {
        dispatch(updateTodos(todos));
    }

    function getAll() {
        const todos = getTodoList();
        set(todos);
    }

    function getFinished() {
        const todos = getFinishedTodos();
        set(todos);
    }

    function getUnfinished() {
        const todos = getUnfinishedTodos();
        set(todos);
    }

    function setFiltered() {
        if(filter === FilterType.FINISHED) {
            getFinished();
        }
        else if(filter === FilterType.UNFINISHED) {
            getUnfinished();
        }
        else {
            getAll();
        }
    }

    function add(text: string, isActive: boolean) {
        const todoList = getTodoList();
        const todo: ITodoItem = {
            id: new Date().toISOString(),
            description: text,
            checked: isActive
        };
        todoList.push(todo);
        saveTodoList(todoList);
        setFiltered();
    }

    function switchStatus(todo: ITodoItem) {
        switchTodo(todo);
        setFiltered();
    }

    function remove(todo: ITodoItem) {
        let todoList = getTodoList();
        todoList = todoList.filter((item) => item.id !== todo.id);
        saveTodoList(todoList);
        dispatch(removeTodo(todo));
    }

    function clearFinished() {
        const todoList = removeFinishedTodos();
        set(todoList);
    }

    return {
        set,
        getAll,
        getFinished,
        getUnfinished,
        setFiltered,
        add,
        switchStatus,
        remove,
        clearFinished
    }
}