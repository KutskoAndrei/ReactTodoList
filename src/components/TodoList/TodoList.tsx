import ITodoItem from '../../models/ITodoItem';
import { useAppSelector } from '../../hooks/useAppSelector';
import classes from './TodoLilst.module.scss';
import TodoItem from '../TodoItem/TodoItem';
import { useEffect } from 'react';
import { useTodos } from '../../hooks/useTodos';
import { useTodoRepository } from '../../hooks/useTodoRepository';

const TodoList = () => {
    const todos: ITodoItem[] = useAppSelector((state) => state.todos.todos);
    const { set } = useTodos();
    const { getTodoList } = useTodoRepository();

    useEffect(() => {
        const initTodoList = getTodoList();
        set(initTodoList);
    }, []);

    return (
        <ul className={classes.todoList} data-testid="todoList">
            {todos && todos.map((todo: ITodoItem) =>
                <li>
                    <TodoItem key={todo.id} todoItem={todo} />
                </li>
            )}
        </ul>
    )
}

export default TodoList;