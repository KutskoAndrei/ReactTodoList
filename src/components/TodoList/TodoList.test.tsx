import ITodoItem from "../../models/ITodoItem";
import TodoList from "./TodoList";
import { RenderResult, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../store/store"; 
import { useTodoRepository } from "../../hooks/useTodoRepository";

const mockItem: ITodoItem[] = [
    {
        id: 'testId',
        description: 'testDescription',
        checked: true
    },
    {
        id: 'testId2',
        description: 'testDescription2',
        checked: false
    },
    {
        id: 'testId3',
        description: 'testDescription3',
        checked: true
    }
];

describe('TodoList tests', () => {
    const { saveTodoList } = useTodoRepository();
    let component: RenderResult;

    saveTodoList(mockItem);
    beforeEach(() => {
        component = render(
            <Provider store={store}>
                <TodoList />
            </Provider>
        );
    });

    it('should match snapshot', () => {
        expect(component).toMatchSnapshot();
    });

    it('should render component with 3 todos', () => {
        expect(screen.queryAllByText(/testDescription/).length).toBe(3);
    });
});