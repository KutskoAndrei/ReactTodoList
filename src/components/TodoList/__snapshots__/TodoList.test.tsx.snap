// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList tests should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="todoList"
      >
        <div
          class="todoItem"
        >
          <div
            class="checkedTodo"
          >
            <div
              class="checkboxItem"
            >
              <input
                checked=""
                type="checkbox"
              />
            </div>
            <h3>
              testDescription
            </h3>
          </div>
          <div
            class="removeBtn"
          >
            <svg
              fill="currentColor"
              fill-rule="evenodd"
              height="30"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="30"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M799.855 166.312c.023.007.043.018.084.059l57.69 57.69c.041.041.052.06.059.084a.118.118 0 0 1 0 .069c-.007.023-.018.042-.059.083L569.926 512l287.703 287.703c.041.04.052.06.059.083a.118.118 0 0 1 0 .07c-.007.022-.018.042-.059.083l-57.69 57.69c-.041.041-.06.052-.084.059a.118.118 0 0 1-.069 0c-.023-.007-.042-.018-.083-.059L512 569.926 224.297 857.629c-.04.041-.06.052-.083.059a.118.118 0 0 1-.07 0c-.022-.007-.042-.018-.083-.059l-57.69-57.69c-.041-.041-.052-.06-.059-.084a.118.118 0 0 1 0-.069c.007-.023.018-.042.059-.083L454.073 512 166.371 224.297c-.041-.04-.052-.06-.059-.083a.118.118 0 0 1 0-.07c.007-.022.018-.042.059-.083l57.69-57.69c.041-.041.06-.052.084-.059a.118.118 0 0 1 .069 0c.023.007.042.018.083.059L512 454.073l287.703-287.702c.04-.041.06-.052.083-.059a.118.118 0 0 1 .07 0Z"
              />
            </svg>
          </div>
        </div>
        <div
          class="todoItem"
        >
          <div
            class="uncheckedTodo"
          >
            <div
              class="checkboxItem"
            >
              <input
                type="checkbox"
              />
            </div>
            <h3>
              testDescription2
            </h3>
          </div>
          <div
            class="removeBtn"
          >
            <svg
              fill="currentColor"
              fill-rule="evenodd"
              height="30"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="30"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M799.855 166.312c.023.007.043.018.084.059l57.69 57.69c.041.041.052.06.059.084a.118.118 0 0 1 0 .069c-.007.023-.018.042-.059.083L569.926 512l287.703 287.703c.041.04.052.06.059.083a.118.118 0 0 1 0 .07c-.007.022-.018.042-.059.083l-57.69 57.69c-.041.041-.06.052-.084.059a.118.118 0 0 1-.069 0c-.023-.007-.042-.018-.083-.059L512 569.926 224.297 857.629c-.04.041-.06.052-.083.059a.118.118 0 0 1-.07 0c-.022-.007-.042-.018-.083-.059l-57.69-57.69c-.041-.041-.052-.06-.059-.084a.118.118 0 0 1 0-.069c.007-.023.018-.042.059-.083L454.073 512 166.371 224.297c-.041-.04-.052-.06-.059-.083a.118.118 0 0 1 0-.07c.007-.022.018-.042.059-.083l57.69-57.69c.041-.041.06-.052.084-.059a.118.118 0 0 1 .069 0c.023.007.042.018.083.059L512 454.073l287.703-287.702c.04-.041.06-.052.083-.059a.118.118 0 0 1 .07 0Z"
              />
            </svg>
          </div>
        </div>
        <div
          class="todoItem"
        >
          <div
            class="checkedTodo"
          >
            <div
              class="checkboxItem"
            >
              <input
                checked=""
                type="checkbox"
              />
            </div>
            <h3>
              testDescription3
            </h3>
          </div>
          <div
            class="removeBtn"
          >
            <svg
              fill="currentColor"
              fill-rule="evenodd"
              height="30"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="30"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M799.855 166.312c.023.007.043.018.084.059l57.69 57.69c.041.041.052.06.059.084a.118.118 0 0 1 0 .069c-.007.023-.018.042-.059.083L569.926 512l287.703 287.703c.041.04.052.06.059.083a.118.118 0 0 1 0 .07c-.007.022-.018.042-.059.083l-57.69 57.69c-.041.041-.06.052-.084.059a.118.118 0 0 1-.069 0c-.023-.007-.042-.018-.083-.059L512 569.926 224.297 857.629c-.04.041-.06.052-.083.059a.118.118 0 0 1-.07 0c-.022-.007-.042-.018-.083-.059l-57.69-57.69c-.041-.041-.052-.06-.059-.084a.118.118 0 0 1 0-.069c.007-.023.018-.042.059-.083L454.073 512 166.371 224.297c-.041-.04-.052-.06-.059-.083a.118.118 0 0 1 0-.07c.007-.022.018-.042.059-.083l57.69-57.69c.041-.041.06-.052.084-.059a.118.118 0 0 1 .069 0c.023.007.042.018.083.059L512 454.073l287.703-287.702c.04-.041.06-.052.083-.059a.118.118 0 0 1 .07 0Z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="todoList"
    >
      <div
        class="todoItem"
      >
        <div
          class="checkedTodo"
        >
          <div
            class="checkboxItem"
          >
            <input
              checked=""
              type="checkbox"
            />
          </div>
          <h3>
            testDescription
          </h3>
        </div>
        <div
          class="removeBtn"
        >
          <svg
            fill="currentColor"
            fill-rule="evenodd"
            height="30"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="30"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M799.855 166.312c.023.007.043.018.084.059l57.69 57.69c.041.041.052.06.059.084a.118.118 0 0 1 0 .069c-.007.023-.018.042-.059.083L569.926 512l287.703 287.703c.041.04.052.06.059.083a.118.118 0 0 1 0 .07c-.007.022-.018.042-.059.083l-57.69 57.69c-.041.041-.06.052-.084.059a.118.118 0 0 1-.069 0c-.023-.007-.042-.018-.083-.059L512 569.926 224.297 857.629c-.04.041-.06.052-.083.059a.118.118 0 0 1-.07 0c-.022-.007-.042-.018-.083-.059l-57.69-57.69c-.041-.041-.052-.06-.059-.084a.118.118 0 0 1 0-.069c.007-.023.018-.042.059-.083L454.073 512 166.371 224.297c-.041-.04-.052-.06-.059-.083a.118.118 0 0 1 0-.07c.007-.022.018-.042.059-.083l57.69-57.69c.041-.041.06-.052.084-.059a.118.118 0 0 1 .069 0c.023.007.042.018.083.059L512 454.073l287.703-287.702c.04-.041.06-.052.083-.059a.118.118 0 0 1 .07 0Z"
            />
          </svg>
        </div>
      </div>
      <div
        class="todoItem"
      >
        <div
          class="uncheckedTodo"
        >
          <div
            class="checkboxItem"
          >
            <input
              type="checkbox"
            />
          </div>
          <h3>
            testDescription2
          </h3>
        </div>
        <div
          class="removeBtn"
        >
          <svg
            fill="currentColor"
            fill-rule="evenodd"
            height="30"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="30"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M799.855 166.312c.023.007.043.018.084.059l57.69 57.69c.041.041.052.06.059.084a.118.118 0 0 1 0 .069c-.007.023-.018.042-.059.083L569.926 512l287.703 287.703c.041.04.052.06.059.083a.118.118 0 0 1 0 .07c-.007.022-.018.042-.059.083l-57.69 57.69c-.041.041-.06.052-.084.059a.118.118 0 0 1-.069 0c-.023-.007-.042-.018-.083-.059L512 569.926 224.297 857.629c-.04.041-.06.052-.083.059a.118.118 0 0 1-.07 0c-.022-.007-.042-.018-.083-.059l-57.69-57.69c-.041-.041-.052-.06-.059-.084a.118.118 0 0 1 0-.069c.007-.023.018-.042.059-.083L454.073 512 166.371 224.297c-.041-.04-.052-.06-.059-.083a.118.118 0 0 1 0-.07c.007-.022.018-.042.059-.083l57.69-57.69c.041-.041.06-.052.084-.059a.118.118 0 0 1 .069 0c.023.007.042.018.083.059L512 454.073l287.703-287.702c.04-.041.06-.052.083-.059a.118.118 0 0 1 .07 0Z"
            />
          </svg>
        </div>
      </div>
      <div
        class="todoItem"
      >
        <div
          class="checkedTodo"
        >
          <div
            class="checkboxItem"
          >
            <input
              checked=""
              type="checkbox"
            />
          </div>
          <h3>
            testDescription3
          </h3>
        </div>
        <div
          class="removeBtn"
        >
          <svg
            fill="currentColor"
            fill-rule="evenodd"
            height="30"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="30"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M799.855 166.312c.023.007.043.018.084.059l57.69 57.69c.041.041.052.06.059.084a.118.118 0 0 1 0 .069c-.007.023-.018.042-.059.083L569.926 512l287.703 287.703c.041.04.052.06.059.083a.118.118 0 0 1 0 .07c-.007.022-.018.042-.059.083l-57.69 57.69c-.041.041-.06.052-.084.059a.118.118 0 0 1-.069 0c-.023-.007-.042-.018-.083-.059L512 569.926 224.297 857.629c-.04.041-.06.052-.083.059a.118.118 0 0 1-.07 0c-.022-.007-.042-.018-.083-.059l-57.69-57.69c-.041-.041-.052-.06-.059-.084a.118.118 0 0 1 0-.069c.007-.023.018-.042.059-.083L454.073 512 166.371 224.297c-.041-.04-.052-.06-.059-.083a.118.118 0 0 1 0-.07c.007-.022.018-.042.059-.083l57.69-57.69c.041-.041.06-.052.084-.059a.118.118 0 0 1 .069 0c.023.007.042.018.083.059L512 454.073l287.703-287.702c.04-.041.06-.052.083-.059a.118.118 0 0 1 .07 0Z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
